drv/cosmo-hf/src/main.rs

```diff
@@ -0,0 +1,309 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Minimal driver for FMC-attached NOR flash, implementing the `hf` API
+//!
+//! The NOR flash chip is a W25Q01JVZEIQ, which is a 1 GiB NOR flash.  It is
+//! connected to the FPGA over SPI / QSPI.
+//!
+//! # References
+//! - [SPI NOR controller](https://github.com/oxidecomputer/quartz/tree/main/hdl/ip/vhd/spi_nor_controller/docs)
+
+#![no_std]
+#![no_main]
+
+use userlib::hl::sleep_for;
+
+use ringbuf::{counted_ringbuf, ringbuf_entry};
+
+mod hf; // implementation of `HostFlash` API
+
+#[derive(Debug, Clone, Copy, PartialEq, counters::Count)]
+enum Trace {
+    #[count(skip)]
+    None,
+
+    FpgaBusy(u32),
+    SectorEraseBusy,
+    WriteBusy,
+
+    HashUpdateError(drv_hash_api::HashError),
+    HashFinalizeError(drv_hash_api::HashError),
```

Comment:
do we care about having counters for the individual `HashError` variants here or not? i'll note that we probably generate IPC counters for them already, so may not need to count them here as well...

Reply:
I don't feel strongly either way, but my vibe is that the IPC counters are sufficient for per-error counting.

Reply:
I think so too, let's stick with that.